tools_lib/src/stack_collapse.rs

```diff
@@ -1,183 +1,215 @@
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::time::{Duration, SystemTime};
+use std::convert::TryInto;
 
-use measureme::Event;
+use measureme::{Event, ProfilingData};
 
-pub fn collapse_stacks<'a>(
-    events: impl Iterator<Item = Event<'a>>,
-    interval: u64,
-) -> HashMap<String, usize> {
-    let mut recorded_stacks = HashMap::<String, usize>::new();
-    let mut thread_stacks: HashMap<u64, (SystemTime, Vec<Event>)> = HashMap::new();
-
-    for event in events {
-        let (next_observation_time, thread_stack) = thread_stacks
+
+fn find_all_thread_ids_and_start_times(profiling_data: &ProfilingData) -> HashMap<u64, SystemTime> {
+
+    let mut thread_ids = HashMap::new();
+
+    for event in profiling_data.iter().filter(|e| !e.timestamp.is_instant()) {
+        thread_ids
             .entry(event.thread_id)
-            .or_insert((event.timestamp, Vec::new()));
-        //if this event is after the next_observation_time then we need to record the current stacks
-        if event.timestamp > *next_observation_time {
-            let mut stack_string = String::new();
-            stack_string.push_str("rustc");
-
-            for event in thread_stack.iter() {
-                stack_string.push(';');
-                stack_string.push_str(&event.label);
-            }
+            .and_modify(|start| {
+                *start = std::cmp::min(event.timestamp.start(), *start);
+            })
+            .or_insert(event.timestamp.start());
+    }
+
+    thread_ids
+}
 
-            let count = recorded_stacks.entry(stack_string).or_default();
+fn process_thread(
+    thread_id: u64,
+    thread_start_time: SystemTime,
+    profiling_data: &ProfilingData,
+    sampling_interval: Duration,
+    counters: &mut HashMap<String, u64>,
+) {
+
+    let mut events = profiling_data
+        .iter()
+        .rev()
+        .filter(|e| !e.timestamp.is_instant())
+        .filter(|e| e.thread_id == thread_id)
+        .peekable();
+
+    let mut sample_cursor = if let Some(last_event) = events.peek() {
+        // Start sample (backwards) at the last position that we would hit if
+        // we sampled forward starting from `thread_start_time`.
+
+        let max_sampled_timestamp = last_event.timestamp.end() - Duration::from_nanos(1);
+
+        if max_sampled_timestamp < thread_start_time {
+            // Just ignore this zero-length thread
+            return
+        }
 
-            while event.timestamp > *next_observation_time {
-                *count += 1;
-                *next_observation_time += Duration::from_millis(interval);
+        let sampling_interval_nanos = sampling_interval.as_nanos();
+
+        let thread_duration_nanos = max_sampled_timestamp.duration_since(thread_start_time).unwrap().as_nanos();
+        let samples = thread_duration_nanos / sampling_interval_nanos;
+        thread_start_time + Duration::from_nanos((samples * sampling_interval_nanos).try_into().unwrap())
```

Comment:
Doing all this logic here might be overkill. We could also start sampling at 
- `last_event.end()`, which would mean there is always one sampling point with stack `rustc`, or
- `last_event.end() - Duration::from_nanos(1)`, which would mean that the earliest sampling point is not exactly at the beginning of the earliest event.

Either case should be fine for real world profiling data. Thoughts?

Reply:
think that is is overkill to track the thread start so in this case I prefere `last_event.end() - Duration::from_nanos(1)`